File: lib/slatedb/reader.rb
# frozen_string_literal: true

module SlateDb
  class Reader
    class << self
      # Open a read-only reader at the given path.
      #
      # @param path [String] The path identifier for the database
      # @param url [String, nil] Optional object store URL
      # @param checkpoint_id [String, nil] Optional checkpoint UUID to read at
      # @param manifest_poll_interval [Integer, nil] Poll interval in milliseconds
      # @param checkpoint_lifetime [Integer, nil] Checkpoint lifetime in milliseconds
      # @param max_memtable_bytes [Integer, nil] Maximum memtable size in bytes
      # @param cache_root [String, nil] Root folder for the reader's local on-disk
      #   object-store cache. Setting this enables the cached object store; when it is
      #   not set the cache (and `max_open_file_handles`) has no effect.
      # @param max_open_file_handles [Integer, nil] Maximum number of file handles to keep
      #   open in the reader's file-handle cache. When the limit is reached, the least
      #   recently used handle is closed (default: 1000). Only takes effect when
      #   `cache_root` is set. (Requires SlateDB >= 0.13.0)
      # @param merge_operator [Symbol, String, nil] Optional merge operator ("string_concat" or "concat")
      # @yield [reader] If a block is given, yields the reader and ensures it's closed
      # @return [Reader] The opened reader (or block result if block given)
      #
      # @example Open a reader
      #   reader = SlateDb::Reader.open("/tmp/mydb")
      #   value = reader.get("key")
      #   reader.close
      #
      # @example Open with block (auto-close)
      #   SlateDb::Reader.open("/tmp/mydb") do |reader|
      #     reader.get("key")
      #   end # automatically closed
      #
      # @example Open at a specific checkpoint
      #   reader = SlateDb::Reader.open("/tmp/mydb", checkpoint_id: "uuid-here")
      #
      # @example Enable the on-disk cache and cap its open file handles
      #   reader = SlateDb::Reader.open("/tmp/mydb",
      #                                 cache_root: "/var/cache/slatedb",
      #                                 max_open_file_handles: 256)
      #
      def open(path, url: nil, checkpoint_id: nil,
               manifest_poll_interval: nil, checkpoint_lifetime: nil,
               max_memtable_bytes: nil, cache_root: nil, max_open_file_handles: nil,
               merge_operator: nil)
        opts = {}
        opts[:manifest_poll_interval] = manifest_poll_interval if manifest_poll_interval
        opts[:checkpoint_lifetime] = checkpoint_lifetime if checkpoint_lifetime
        opts[:max_memtable_bytes] = max_memtable_bytes if max_memtable_bytes
        opts[:cache_root] = cache_root if cache_root
        opts[:max_open_file_handles] = max_open_file_handles if max_open_file_handles
        opts[:merge_operator] = merge_operator.to_s if merge_operator

        reader = _open(path, url, checkpoint_id, opts)

        if block_given?
          begin
            yield reader
          ensure
            begin
              reader.close
            rescue StandardError
              nil
            end
          end
        else
          reader
        end
      end
    end

    # Get a value by key.
    #
    # @param key [String] The key to look up
    # @param durability_filter [String, nil] Filter by durability level ("remote" or "memory")
    # @param dirty [Boolean, nil] Whether to include uncommitted data
    # @param cache_blocks [Boolean, nil] Whether to cache blocks
    # @return [String, nil] The value, or nil if not found
    #
    def get(key, durability_filter: nil, dirty: nil, cache_blocks: nil)
      opts = {}
      opts[:durability_filter] = durability_filter.to_s if durability_filter
      opts[:dirty] = dirty unless dirty.nil?
      opts[:cache_blocks] = cache_blocks unless cache_blocks.nil?

      if opts.empty?
        _get(key)
      else
        _get_with_options(key, opts)
      end
    end

    # Scan a range of keys.
    #
    # @param start_key [String] The start key (inclusive)
    # @param end_key [String, nil] The end key (exclusive)
    # @return [Iterator] An iterator over key-value pairs
    #
    def scan(start_key, end_key = nil, durability_filter: nil, dirty: nil,
             read_ahead_bytes: nil, cache_blocks: nil, max_fetch_tasks: nil, &)
      opts = {}
      opts[:durability_filter] = durability_filter.to_s if durability_filter
      opts[:dirty] = dirty unless dirty.nil?
      opts[:read_ahead_bytes] = read_ahead_bytes if read_ahead_bytes
      opts[:cache_blocks] = cache_blocks unless cache_blocks.nil?
      opts[:max_fetch_tasks] = max_fetch_tasks if max_fetch_tasks

      iter = if opts.empty?
               _scan(start_key, end_key)
             else
               _scan_with_options(start_key, end_key, opts)
             end

      if block_given?
        iter.each(&)
      else
        iter
      end
    end

    # Scan all keys with a given prefix.
    #
    # @param prefix [String] The key prefix to scan
    # @param durability_filter [String, nil] Filter by durability level
    # @param dirty [Boolean, nil] Whether to include uncommitted data
    # @param read_ahead_bytes [Integer, nil] Number of bytes to read ahead
    # @param cache_blocks [Boolean, nil] Whether to cache blocks
    # @param max_fetch_tasks [Integer, nil] Maximum number of fetch tasks
    # @param order [Symbol, String, nil] Iteration order (:asc/:ascending or :desc/:descending)
    # @param suffix [Range, nil] Restrict the scan to a range of key *suffixes*
    #   (the part of each key after +prefix+). See {Database#scan_prefix} for
    #   the semantics. (Requires SlateDB >= 0.14.0)
    # @return [Iterator] An iterator over key-value pairs
    #
    def scan_prefix(prefix, durability_filter: nil, dirty: nil,
                    read_ahead_bytes: nil, cache_blocks: nil, max_fetch_tasks: nil,
                    order: nil, suffix: nil, &)
      opts = SlateDb.scan_options(
        durability_filter: durability_filter, dirty: dirty, read_ahead_bytes: read_ahead_bytes,
        cache_blocks: cache_blocks, max_fetch_tasks: max_fetch_tasks, order: order
      )
      opts.merge!(SlateDb.suffix_range_options(suffix))

      iter = if opts.empty?
               _scan_prefix(prefix)
             else
               _scan_prefix_with_options(prefix, opts)
             end

      if block_given?
        iter.each(&)
      else
        iter
      end
    end
  end
end

File: ext/slatedb/src/reader.rs
use std::sync::Arc;

use magnus::prelude::*;
use magnus::{function, method, Error, RHash, Ruby};
use slatedb::config::{DbReaderOptions, DurabilityLevel, ReadOptions, ScanOptions};
use slatedb::DbReader;
use slatedb::IterationOrder;

use crate::errors::invalid_argument_error;
use crate::iterator::Iterator;
use crate::runtime::block_on_result;
use crate::utils::{get_optional, prefix_subrange_from_kwargs, resolve_object_store};

/// Ruby wrapper for SlateDB Reader.
///
/// This struct is exposed to Ruby as `SlateDb::Reader`.
/// Provides read-only access to a database, optionally pinned to a checkpoint.
#[magnus::wrap(class = "SlateDb::Reader", free_immediately, size)]
pub struct Reader {
    inner: Arc<DbReader>,
}

impl Reader {
    /// Open a reader at the given path.
    ///
    /// # Arguments
    /// * `path` - The path identifier for the database
    /// * `url` - Optional object store URL
    /// * `checkpoint_id` - Optional checkpoint UUID to read at
    /// * `kwargs` - Additional options (manifest_poll_interval, checkpoint_lifetime,
    ///   max_memtable_bytes, skip_wal_replay, cache_root, max_open_file_handles).
    ///   The local disk cache (and therefore `max_open_file_handles`) is only active
    ///   when `cache_root` is set.
    pub fn open(
        path: String,
        url: Option<String>,
        checkpoint_id: Option<String>,
        kwargs: RHash,
    ) -> Result<Self, Error> {
        // Parse options
        let manifest_poll_interval = get_optional::<u64>(&kwargs, "manifest_poll_interval")?
            .map(std::time::Duration::from_millis);
        let checkpoint_lifetime = get_optional::<u64>(&kwargs, "checkpoint_lifetime")?
            .map(std::time::Duration::from_millis);
        let max_memtable_bytes = get_optional::<u64>(&kwargs, "max_memtable_bytes")?;
        let skip_wal_replay = get_optional::<bool>(&kwargs, "skip_wal_replay")?;
        let max_open_file_handles = get_optional::<usize>(&kwargs, "max_open_file_handles")?;
        let cache_root = get_optional::<String>(&kwargs, "cache_root")?;

        // Parse checkpoint_id as UUID
        let checkpoint_uuid =
            if let Some(id_str) = checkpoint_id {
                Some(uuid::Uuid::parse_str(&id_str).map_err(|e| {
                    invalid_argument_error(&format!("invalid checkpoint_id: {}", e))
                })?)
            } else {
                None
            };

        let reader = block_on_result(async {
            let object_store: Arc<dyn slatedb::object_store::ObjectStore> =
                if let Some(ref url) = url {
                    resolve_object_store(url)?
                } else {
                    Arc::new(slatedb::object_store::memory::InMemory::new())
                };

            let mut options = DbReaderOptions::default();
            if let Some(interval) = manifest_poll_interval {
                options.manifest_poll_interval = interval;
            }
            if let Some(lifetime) = checkpoint_lifetime {
                options.checkpoint_lifetime = lifetime;
            }
            if let Some(max_bytes) = max_memtable_bytes {
                options.max_memtable_bytes = max_bytes;
            }
            if let Some(skip_replay) = skip_wal_replay {
                options.skip_wal_replay = skip_replay;
            }
            if let Some(ref root) = cache_root {
                options.object_store_cache_options.root_folder =
                    Some(std::path::PathBuf::from(root));
            }
            if let Some(max_handles) = max_open_file_handles {
                options.object_store_cache_options.max_open_file_handles = max_handles;
            }
            DbReader::open(path, object_store, checkpoint_uuid, options).await
        })?;

        Ok(Self {
            inner: Arc::new(reader),
        })
    }

    /// Get a value by key.
    pub fn get(&self, key: String) -> Result<Option<String>, Error> {
        if key.is_empty() {
            return Err(invalid_argument_error("key cannot be empty"));
        }

        let result = block_on_result(async { self.inner.get(key.as_bytes()).await })?;
        Ok(result.map(|b| String::from_utf8_lossy(&b).to_string()))
    }

    /// Get a value by key with options.
    pub fn get_with_options(&self, key: String, kwargs: RHash) -> Result<Option<String>, Error> {
        if key.is_empty() {
            return Err(invalid_argument_error("key cannot be empty"));
        }

        let mut opts = ReadOptions::default();

        if let Some(df) = get_optional::<String>(&kwargs, "durability_filter")? {
            opts.durability_filter = match df.as_str() {
                "remote" => DurabilityLevel::Remote,
                "memory" => DurabilityLevel::Memory,
                other => {
                    return Err(invalid_argument_error(&format!(
                        "invalid durability_filter: {} (expected 'remote' or 'memory')",
                        other
                    )))
                }
            };
        }

        if let Some(dirty) = get_optional::<bool>(&kwargs, "dirty")? {
            opts.dirty = dirty;
        }

        if let Some(cb) = get_optional::<bool>(&kwargs, "cache_blocks")? {
            opts.cache_blocks = cb;
        }

        let result =
            block_on_result(async { self.inner.get_with_options(key.as_bytes(), &opts).await })?;
        Ok(result.map(|b| String::from_utf8_lossy(&b).to_string()))
    }

    /// Get a value by key as raw bytes.
    pub fn get_bytes(&self, key: String) -> Result<Option<Vec<u8>>, Error> {
        if key.is_empty() {
            return Err(invalid_argument_error("key cannot be empty"));
        }

        let result = block_on_result(async { self.inner.get(key.as_bytes()).await })?;
        Ok(result.map(|b| b.to_vec()))
    }

    /// Scan a range of keys.
    pub fn scan(&self, start: String, end_key: Option<String>) -> Result<Iterator, Error> {
        if start.is_empty() {
            return Err(invalid_argument_error("start key cannot be empty"));
        }

        let start_bytes = start.into_bytes();
        let end_bytes = end_key.map(|e| e.into_bytes());

        let iter = block_on_result(async {
            match end_bytes {
                Some(end) => self.inner.scan(start_bytes..end).await,
                None => self.inner.scan(start_bytes..).await,
            }
        })?;

        Ok(Iterator::new(iter))
    }

    /// Scan a range of keys with options.
    pub fn scan_with_options(
        &self,
        start: String,
        end_key: Option<String>,
        kwargs: RHash,
    ) -> Result<Iterator, Error> {
        if start.is_empty() {
            return Err(invalid_argument_error("start key cannot be empty"));
        }

        let mut opts = ScanOptions::default();

        if let Some(df) = get_optional::<String>(&kwargs, "durability_filter")? {
            opts.durability_filter = match df.as_str() {
                "remote" => DurabilityLevel::Remote,
                "memory" => DurabilityLevel::Memory,
                other => {
                    return Err(invalid_argument_error(&format!(
                        "invalid durability_filter: {} (expected 'remote' or 'memory')",
                        other
                    )))
                }
            };
        }

        if let Some(dirty) = get_optional::<bool>(&kwargs, "dirty")? {
            opts.dirty = dirty;
        }

        if let Some(rab) = get_optional::<usize>(&kwargs, "read_ahead_bytes")? {
            opts.read_ahead_bytes = rab;
        }

        if let Some(cb) = get_optional::<bool>(&kwargs, "cache_blocks")? {
            opts.cache_blocks = cb;
        }

        if let Some(mft) = get_optional::<usize>(&kwargs, "max_fetch_tasks")? {
            opts.max_fetch_tasks = mft;
        }
        if let Some(order) = get_optional::<String>(&kwargs, "order")? {
            opts.order = match order.as_str() {
                "ascending" | "asc" => IterationOrder::Ascending,
                "descending" | "desc" => IterationOrder::Descending,
                other => {
                    return Err(invalid_argument_error(&format!(
                        "invalid order: {} (expected 'asc' or 'desc')",
                        other
                    )))
                }
            };
        }

        let start_bytes = start.into_bytes();
        let end_bytes = end_key.map(|e| e.into_bytes());

        let iter = block_on_result(async {
            match end_bytes {
                Some(end) => self.inner.scan_with_options(start_bytes..end, &opts).await,
                None => self.inner.scan_with_options(start_bytes.., &opts).await,
            }
        })?;

        Ok(Iterator::new(iter))
    }

    /// Scan all keys with a given prefix.
    pub fn scan_prefix(&self, prefix: String) -> Result<Iterator, Error> {
        if prefix.is_empty() {
            return Err(invalid_argument_error("prefix cannot be empty"));
        }

        let iter =
            block_on_result(async { self.inner.scan_prefix(prefix.as_bytes(), ..).await })?;

        Ok(Iterator::new(iter))
    }

    /// Scan all keys with a given prefix with options.
    pub fn scan_prefix_with_options(
        &self,
        prefix: String,
        kwargs: RHash,
    ) -> Result<Iterator, Error> {
        if prefix.is_empty() {
            return Err(invalid_argument_error("prefix cannot be empty"));
        }

        let mut opts = ScanOptions::default();

        if let Some(df) = get_optional::<String>(&kwargs, "durability_filter")? {
            opts.durability_filter = match df.as_str() {
                "remote" => DurabilityLevel::Remote,
                "memory" => DurabilityLevel::Memory,
                other => {
                    return Err(invalid_argument_error(&format!(
                        "invalid durability_filter: {} (expected 'remote' or 'memory')",
                        other
                    )))
                }
            };
        }

        if let Some(dirty) = get_optional::<bool>(&kwargs, "dirty")? {
            opts.dirty = dirty;
        }

        if let Some(rab) = get_optional::<usize>(&kwargs, "read_ahead_bytes")? {
            opts.read_ahead_bytes = rab;
        }

        if let Some(cb) = get_optional::<bool>(&kwargs, "cache_blocks")? {
            opts.cache_blocks = cb;
        }

        if let Some(mft) = get_optional::<usize>(&kwargs, "max_fetch_tasks")? {
            opts.max_fetch_tasks = mft;
        }
        if let Some(order) = get_optional::<String>(&kwargs, "order")? {
            opts.order = match order.as_str() {
                "ascending" | "asc" => IterationOrder::Ascending,
                "descending" | "desc" => IterationOrder::Descending,
                other => {
                    return Err(invalid_argument_error(&format!(
                        "invalid order: {} (expected 'asc' or 'desc')",
                        other
                    )))
                }
            };
        }

        let subrange = prefix_subrange_from_kwargs(&kwargs)?;
        let iter = block_on_result(async {
            self.inner
                .scan_prefix_with_options(prefix.as_bytes(), subrange, &opts)
                .await
        })?;

        Ok(Iterator::new(iter))
    }

    /// Close the reader.
    pub fn close(&self) -> Result<(), Error> {
        block_on_result(async { self.inner.close().await })?;
        Ok(())
    }
}

/// Define the Reader class on the SlateDb module.
pub fn define_reader_class(ruby: &Ruby, module: &magnus::RModule) -> Result<(), Error> {
    let class = module.define_class("Reader", ruby.class_object())?;

    // Class methods
    class.define_singleton_method("_open", function!(Reader::open, 4))?;

    // Instance methods
    class.define_method("_get", method!(Reader::get, 1))?;
    class.define_method("_get_with_options", method!(Reader::get_with_options, 2))?;
    class.define_method("get_bytes", method!(Reader::get_bytes, 1))?;
    class.define_method("_scan", method!(Reader::scan, 2))?;
    class.define_method("_scan_with_options", method!(Reader::scan_with_options, 3))?;
    class.define_method("_scan_prefix", method!(Reader::scan_prefix, 1))?;
    class.define_method(
        "_scan_prefix_with_options",
        method!(Reader::scan_prefix_with_options, 2),
    )?;
    class.define_method("close", method!(Reader::close, 0))?;

    Ok(())
}
